### ISSUE TITLE
linter: getAllComments() end offset is wrong in a file with BOM

### ISSUE THREAD

**Issue by IvMisticos** [oxc-project/oxc#26162]
### What version of Oxlint are you using?

1.78.0

### What command did you run?

[bunx] oxlint bom.js

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{
  "jsPlugins": ["./plugin.js"],
  "rules": {
    "plugin/rule": "error"
  }
}
```
```ts
export default {
  meta: {
    name: "plugin",
  },
  rules: {
    rule: {
      create: (context) => ({
        Program() {
          const comment = context.sourceCode.getAllComments()[0];
          console.log(
            comment.start,
            comment.end,
            JSON.stringify(context.sourceCode.text.slice(comment.start, comment.end)),
          );
        },
      }),
    },
  },
};
```

### What happened?

When a file 1) starts with BOM 2) has a comment 3) has at least one non-ASCII character following the comment,
Then plugin API gives the wrong end position for a comment.

`context.sourceCode.text`​ does not contain BOM and the length is correct. The comment's end​ from `getAllComments​` is off by one.

Creating example files:
```bash
printf "\xEF\xBB\xBF/*x*/'\xC3\xA9'" > bom.js # bad
printf "/*x*/'\xC3\xA9'" > nobom.js # good
```

**Comment by IvMisticos** [comment]
Likely introduced in #18376

https://github.com/oxc-project/oxc/blob/784e9faad2a5e65e063f3e1f2f448203675909ec/crates/oxc_ast_visit/src/utf8_to_utf16/mod.rs#L53-L56
https://github.com/oxc-project/oxc/blob/784e9faad2a5e65e063f3e1f2f448203675909ec/crates/oxc_ast_visit/src/utf8_to_utf16/translation.rs#L94-L98

<details><summary>AI-suggested fix (im so sorry........)</summary>
<p>

```diff
--- a/crates/oxc_ast_visit/src/utf8_to_utf16/translation.rs
+++ b/crates/oxc_ast_visit/src/utf8_to_utf16/translation.rs
@@ -94,7 +94,7 @@
                 let bytes_in_char =
                     difference_for_this_byte as usize + usize::from(byte >= 0xF0) + 1;
                 #[expect(clippy::cast_possible_truncation)]
-                let utf8_offset = (start_offset + index + bytes_in_char) as u32;
+                let utf8_offset = (start_offset + index + bytes_in_char) as u32 + offset;
                 translations.push(Translation { utf8_offset, utf16_difference });
```

</p>
</details>

**Comment by IvMisticos** [comment]
Similar issue in a different place with no BOM it seems, likely to cause trouble in partial sources with non-ASCII:

https://github.com/oxc-project/oxc/blob/784e9faad2a5e65e063f3e1f2f448203675909ec/crates/oxc_linter/src/lib.rs#L753-L754

& `apply_fixes`:

https://github.com/oxc-project/oxc/blob/784e9faad2a5e65e063f3e1f2f448203675909ec/apps/oxlint/src/js_plugins/fix.rs#L30-L37

& this:

https://github.com/oxc-project/oxc/blob/784e9faad2a5e65e063f3e1f2f448203675909ec/crates/oxc_ast_visit/src/utf8_to_utf16/converter.rs#L302-L319

**Comment by IvMisticos** [comment]
Repro for more/similar:

```js
export default {
  meta: {
    name: "plugin",
  },
  rules: {
    rule: {
      create: (context) => ({
        Program(node) {
          console.log(`[${node.start},${node.end}]`);
        },
      }),
    },
  },
};
```
```bash
# expected [0,4], produces [-1,4]. BOM needed
printf '\xEF\xBB\xBFé=1;' > neg.js

# expected [4,5] (`é`), produces [3,5] (` é`). crlf needed
printf '<template><p>x</p></template>\r\n<script>\r\nlet é = 1;\r\n</script>\r\n' > crlf.vue
```

Fixer is affected by this issue (even though doubt many people would ever hit or notice anything like that), potentially causing a tiny corruption, i.e when using `fixer.replaceTextRange`.

**Comment by saberoueslati** [comment]
I'll be working on this

**Comment by bharadwaj-pendyala** [comment]
Yep, it's the `+ offset` you pointed at. `build_translations` records `utf8_offset` relative to the trimmed source, but `new_with_offset` seeds the table with `{3, 3}`, so every entry after the first non-ASCII character sits 3 bytes early and the comment end lands on the wrong side of it.

On your repro the comment comes back as `0..4` instead of `0..5`. I've got a fix and a regression test, PR on the way.

**Comment by overlookmotel** [comment]
#26173 fixed the bug, and #26223 and #26287 add more test coverage.

I believe this fixes and verifies all the different problems discussed on this thread - they all stemmed from the same source.

Please do re-open this and let us know if that's not the case!
